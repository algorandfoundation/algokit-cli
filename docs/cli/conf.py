# Configuration file for the Sphinx documentation builder.
#
# For the full list of built-in configuration values, see the documentation:
# https://www.sphinx-doc.org/en/master/usage/configuration.html

# -- Project information -----------------------------------------------------
# https://www.sphinx-doc.org/en/master/usage/configuration.html#project-information

project = "AlgoKit"
copyright = "2023, Algorand Foundation"
author = "Algorand Foundation"

# -- General configuration ---------------------------------------------------
# https://www.sphinx-doc.org/en/master/usage/configuration.html#general-configuration

extensions = ["sphinx_click"]

exclude_patterns = ["_build", "Thumbs.db", ".DS_Store"]
smartquotes = False
