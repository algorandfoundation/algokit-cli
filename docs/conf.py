# Configuration file for the Sphinx documentation builder.
#
# For the full list of built-in configuration values, see the documentation:
# https://www.sphinx-doc.org/en/master/usage/configuration.html
import os

# -- Project information -----------------------------------------------------
# https://www.sphinx-doc.org/en/master/usage/configuration.html#project-information

project = "AlgoKit"
copyright = "2023, Algorand Foundation"
author = "Algorand Foundation"

# -- General configuration ---------------------------------------------------
# https://www.sphinx-doc.org/en/master/usage/configuration.html#general-configuration

extensions = ["myst_parser", "sphinx_starlight_builder", "sphinx_markdown_builder"]

exclude_patterns = ["_build", "Thumbs.db", ".DS_Store", "cli/index.rst"]

myst_heading_anchors = 4
markdown_http_base = "."
starlight_http_base = "/starlight"

# TODO: Change based on env
# myst_enable_extensions = [
#     "substitution",
# ]
# myst_substitutions = {
#     "README": "",
#     "ARC1": "",
#     "ARC32": "https://github.com/algorandfoundation/ARCs/blob/main/ARCs/arc-0032.md"
# }
