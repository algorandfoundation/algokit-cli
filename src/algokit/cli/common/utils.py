# Cli/Click related helper functions and classes


import typing as t

import click

from algokit.cli.common.constants import ExplorerEntityType


class MutuallyExclusiveOption(click.Option):
    """
    A Click option that defines mutually exclusive command line options.

    Args:
        *args: Positional arguments passed to the parent class constructor.
        **kwargs: Keyword arguments passed to the parent class constructor.
            not_required_if (list): A list of options that the current option is mutually exclusive with.

    Attributes:
        not_required_if (list): A list of options that the current option is mutually exclusive with.

    Raises:
        AssertionError: If the `not_required_if` parameter is not provided.

    Example:
        ```python
        @click.command()
        @click.option('--option1', help='Option 1')
        @click.option('--option2', help='Option 2')
        @click.option('--option3', help='Option 3', cls=MutuallyExclusiveOption, not_required_if=['option1', 'option2'])
        def my_command(option1, option2, option3):
            # Command logic here
            pass
        ```

        In the example above, the `MutuallyExclusiveOption` class is used to define the `option3` command line option.
        This option is mutually exclusive with `option1` and `option2`,
        meaning that if either `option1` or `option2` is provided, `option3` cannot be used.
        If `option3` is provided along with `option1` or `option2`, a `click.UsageError` is raised.
    """

    def __init__(self, *args: t.Any, **kwargs: t.Any) -> None:
        self.not_required_if: list = kwargs.pop("not_required_if")

        assert self.not_required_if, "'not_required_if' parameter required"
        kwargs["help"] = (
            kwargs.get("help", "") + " Option is mutually exclusive with " + ", ".join(self.not_required_if) + "."
        ).strip()
        super().__init__(*args, **kwargs)

    def handle_parse_result(
        self, ctx: click.Context, opts: t.Mapping[str, t.Any], args: list[str]
    ) -> tuple[t.Any, list[str]]:
        """
        Overrides the `handle_parse_result` method of the `click.Option` class.

        This method checks if the current option is present in the provided options (`opts`)
        and if any of the mutually exclusive options are also present.
        If both the current option and a mutually exclusive option are present, it raises a `click.UsageError`.
        Otherwise, it returns the result of the parent `handle_parse_result` method.

        Args:
            ctx (click.Context): The Click context object.
            opts (dict): The dictionary of parsed options.
            args (list): The list of remaining arguments.

        Returns:
            The result of the parent `handle_parse_result` method.

        Raises:
            click.UsageError: If the current option and a mutually exclusive option are both present.
        """

        current_opt: bool = self.name in opts
        for mutex_opt in self.not_required_if:
            if mutex_opt in opts:
                if current_opt:
                    raise click.UsageError(
                        "Illegal usage: '" + str(self.name) + "' is mutually exclusive with " + str(mutex_opt) + "."
                    )
                self.prompt = None
        return super().handle_parse_result(ctx, opts, args)


def get_explorer_url(identifier: str | int, network: str, entity_type: ExplorerEntityType) -> str:
    """
    Returns a URL for exploring a specified type (transaction, asset, address) on the specified network.

    Args:
        identifier (str | int): The ID of the transaction, asset, or address to explore.
        network (str): The name of the network (e.g., "localnet", "testnet", "mainnet").
        entity_type (ExplorerEntityType): The type to explore (e.g., ExplorerEntityType.TRANSACTION,
        ExplorerEntityType.ASSET, ExplorerEntityType.ADDRESS).

    Returns:
        str: The URL for exploring the specified type on the specified network.

    Raises:
        ValueError: If the network or explorer type is invalid.
    """

    base_urls: dict[str, dict[str, str]] = {
        "testnet": {
            ExplorerEntityType.TRANSACTION.value: "https://testnet.algoexplorer.io/tx/",
            ExplorerEntityType.ASSET.value: "https://testnet.explorer.perawallet.app/assets/",
            ExplorerEntityType.ADDRESS.value: "https://testnet.algoexplorer.io/address/",
        },
        "mainnet": {
            ExplorerEntityType.TRANSACTION.value: "https://algoexplorer.io/tx/",
            ExplorerEntityType.ASSET.value: "https://explorer.perawallet.app/assets/",
            ExplorerEntityType.ADDRESS.value: "https://algoexplorer.io/address/",
        },
    }

    if network == "localnet":
        return f"https://app.dappflow.org/setnetwork?name=sandbox&redirect=explorer/{entity_type.value}/{identifier}/"

    if network not in base_urls:
        raise ValueError(f"Invalid network: {network}")

    if entity_type.value not in base_urls[network]:
        raise ValueError(f"Invalid explorer type: {entity_type}")

    return base_urls[network][entity_type.value] + str(identifier)
