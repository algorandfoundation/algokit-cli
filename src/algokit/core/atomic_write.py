import contextlib
import os
import shutil
import stat
from pathlib import Path
from typing import Literal


def atomic_write(file_contents: str, target_file_path: Path, mode: Literal["a", "w"] = "w") -> None:
    # if target path is a symlink, we want to use the real path as the replacement target,
    # otherwise we'd just be overwriting the symlink
    target_file_path = target_file_path.resolve()
    temp_file_path = target_file_path.with_suffix(f"{target_file_path.suffix}.algokit~")
    try:
        # preserve file metadata if it already exists
        with contextlib.suppress(FileNotFoundError):
            _copy_with_metadata(target_file_path, temp_file_path)
        # write content to new temp file
        with temp_file_path.open(mode=mode, encoding="utf-8") as fp:
            fp.write(file_contents)
        # overwrite destination with the temp file
        temp_file_path.replace(target_file_path)
    finally:
        temp_file_path.unlink(missing_ok=True)


def _copy_with_metadata(source: Path, target: Path) -> None:
    # copy content, stat-info (mode too), timestamps...
    shutil.copy2(source, target)
    # try copy owner+group if platform supports it
    if hasattr(os, "chown"):
        # copy owner and group
        st = source.stat()
        os.chown(target, st[stat.ST_UID], st[stat.ST_GID])
