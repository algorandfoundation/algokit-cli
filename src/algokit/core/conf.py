import logging
import os
import platform
import sys
import typing as t
from importlib import metadata
from pathlib import Path

if sys.version_info >= (3, 11):
    import tomllib
else:
    import tomli as tomllib

PACKAGE_NAME = "algokit"
ALGOKIT_CONFIG = ".algokit.toml"

logger = logging.getLogger(__name__)


def get_app_config_dir() -> Path:
    """Get the application config files location - things that should persist, and potentially follow a user"""
    os_type = platform.system().lower()
    if os_type == "windows":
        config_dir = os.getenv("APPDATA")
    else:
        config_dir = os.getenv("XDG_CONFIG_HOME")
    if config_dir is None:
        config_dir = "~/.config"
    return _get_relative_app_path(config_dir)


def get_app_state_dir() -> Path:
    """Get the application state files location - things the user wouldn't normally interact with directly"""
    os_type = platform.system().lower()
    if os_type == "windows":
        state_dir = os.getenv("LOCALAPPDATA")
    elif os_type == "darwin":
        state_dir = "~/Library/Application Support"
    else:
        state_dir = os.getenv("XDG_STATE_HOME")
    if state_dir is None:
        state_dir = "~/.local/state"
    return _get_relative_app_path(state_dir)


def _get_relative_app_path(base_dir: str) -> Path:
    path = Path(base_dir).expanduser()
    result = path / PACKAGE_NAME
    result.mkdir(parents=True, exist_ok=True)
    # resolve path in case of UWP sandbox redirection
    return result.resolve()


def get_current_package_version() -> str:
    return metadata.version(PACKAGE_NAME)


def get_algokit_config(project_dir: Path) -> dict[str, t.Any] | None:
    """
    Load and parse a TOML configuration file. Will never throw.
    :param project_dir: Project directory path.
    :return: A dictionary containing the configuration or None if not found.
    """
    config_path = project_dir / ALGOKIT_CONFIG
    logger.debug(f"Attempting to load project config from {config_path}")
    try:
        config_text = config_path.read_text("utf-8")
    except FileNotFoundError:
        logger.debug(f"No {ALGOKIT_CONFIG} file found in the project directory.")
        return None
    except Exception as ex:
        logger.debug(f"Unexpected error reading {ALGOKIT_CONFIG} file: {ex}", exc_info=True)
        return None

    try:
        return tomllib.loads(config_text)
    except Exception as ex:
        logger.debug(f"Error parsing {ALGOKIT_CONFIG} file: {ex}", exc_info=True)
        return None
