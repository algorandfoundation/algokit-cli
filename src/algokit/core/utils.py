import re
import socket


def extract_version_triple(version_str: str) -> str:
    match = re.search(r"\d+\.\d+\.\d+", version_str)
    if not match:
        raise ValueError("Unable to parse version number")
    return match.group()


def is_minimum_version(system_version: str, minimum_version: str) -> bool:
    system_version_as_tuple = tuple(map(int, system_version.split(".")))
    minimum_version_as_tuple = tuple(map(int, minimum_version.split(".")))
    return system_version_as_tuple >= minimum_version_as_tuple


def is_network_available(host: str = "8.8.8.8", port: int = 53, timeout: float = 3.0) -> bool:
    """
    Check if internet is available by trying to establish a socket connection.
    """

    try:
        socket.setdefaulttimeout(timeout)
        socket.socket(socket.AF_INET, socket.SOCK_STREAM).connect((host, port))
        return True
    except OSError:
        return False
