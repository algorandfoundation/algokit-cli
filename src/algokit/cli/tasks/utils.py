# AlgoKit Tasks related utility functions

import logging
import os
import stat
import sys

import algosdk
import algosdk.encoding
import click
from algokit_utils import (
    Account,
    get_algod_client,
    get_algonode_config,
    get_default_localnet_config,
)
from algosdk.util import algos_to_microalgos, microalgos_to_algos

from algokit.cli.common.constants import AlgorandNetwork
from algokit.core.tasks.wallet import get_alias

logger = logging.getLogger(__name__)


def _validate_asset_balance(account_info: dict, asset_id: int, decimals: int, amount: int = 0) -> None:
    asset_record = next((asset for asset in account_info.get("assets", []) if asset["asset-id"] == asset_id), None)

    if not asset_record:
        raise click.ClickException("Account is not opted into the asset")

    if amount > 0 and asset_record["amount"] < amount:
        required = amount / 10**decimals
        available = asset_record["amount"] / 10**decimals
        raise click.ClickException(
            f"Insufficient asset balance in account, required: {required}, available: {available}"
        )


def _validate_algo_balance(account_info: dict, amount: int) -> None:
    if account_info.get("amount", 0) < amount:
        required = microalgos_to_algos(amount)  # type: ignore[no-untyped-call]
        available = microalgos_to_algos(account_info.get("amount", 0))  # type: ignore[no-untyped-call]
        raise click.ClickException(
            f"Insufficient Algos balance in account, required: {required} Algos, available: {available} Algos"
        )


def get_private_key_from_mnemonic() -> str:
    """
    Converts a mnemonic phrase into a private key.

    Returns:
        str: The private key generated from the mnemonic phrase.

    Raises:
        click.ClickException: If the entered mnemonic phrase is invalid.

    Example Usage:
        private_key = get_private_key_from_mnemonic()
        print(private_key)

    Inputs:
        None

    Flow:
        1. Prompts the user to enter a mnemonic phrase.
        2. Converts the entered mnemonic phrase into a private key using `algosdk.mnemonic.to_private_key`.
        3. Returns the private key.

    """

    mnemonic_phrase = click.prompt("Enter the mnemonic phrase (25 words separated by whitespace)", hide_input=True)
    try:
        return str(algosdk.mnemonic.to_private_key(mnemonic_phrase))  # type: ignore[no-untyped-call]
    except Exception as err:
        raise click.ClickException("Invalid mnemonic. Please provide a valid Algorand mnemonic.") from err


def load_algod_client(network: AlgorandNetwork) -> algosdk.v2client.algod.AlgodClient:
    """
    Returns an instance of the `algosdk.v2client.algod.AlgodClient` class for the specified network.

    Args:
        network (str): The network for which the `AlgodClient` instance needs to be loaded.

    Returns:
        algosdk.v2client.algod.AlgodClient: An instance of the `AlgodClient` class for the specified network.

    Raises:
        click.ClickException: If the specified network is invalid.
    """

    config_mapping = {
        AlgorandNetwork.LOCALNET: get_default_localnet_config("algod"),
        AlgorandNetwork.TESTNET: get_algonode_config("testnet", "algod", ""),
        AlgorandNetwork.MAINNET: get_algonode_config("mainnet", "algod", ""),
    }
    try:
        return get_algod_client(config_mapping[network])
    except KeyError as err:
        raise click.ClickException("Invalid network") from err


def get_asset_decimals(asset_id: int, algod_client: algosdk.v2client.algod.AlgodClient) -> int:
    """
    Retrieves the number of decimal places for a given asset.

    Args:
        asset_id (int): The ID of the asset for which the number of decimal places is to be retrieved. (0 for Algo)
        algod_client (algosdk.v2client.algod.AlgodClient): An instance of the AlgodClient class.

    Returns:
        int: The number of decimal places for the given asset.

    Raises:
        click.ClickException: If the asset info response is invalid.

    Example:
        asset_id = 123
        algod_client = algosdk.v2client.algod.AlgodClient("https://api.algoexplorer.io", "API_KEY")
        decimals = get_asset_decimals(asset_id, algod_client)
        print(decimals)
    """

    if asset_id == 0:
        return 6

    asset_info = algod_client.asset_info(asset_id)

    if not isinstance(asset_info, dict) or "params" not in asset_info or "decimals" not in asset_info["params"]:
        raise click.ClickException("Invalid asset info response")

    return int(asset_info["params"]["decimals"])


def validate_balance(
    algod_client: algosdk.v2client.algod.AlgodClient, account: Account | str, asset_id: int, amount: int = 0
) -> None:
    """
    Validates the balance of an account before an operation.

    Args:
        algod_client (algosdk.v2client.algod.AlgodClient): The AlgodClient object for
        interacting with the Algorand blockchain.
        account (Account | str): The account object.
        asset_id (int): The ID of the asset to be checked (0 for Algos).
        amount (int): The amount of Algos or asset for the operation. Defaults to 0 implying opt-in check only.

    Raises:
        click.ClickException: If any validation check fails.
    """
    address = account.address if isinstance(account, Account) else account
    account_info = algod_client.account_info(address)

    if not isinstance(account_info, dict):
        raise click.ClickException("Invalid account info response")

    if asset_id == 0:
        _validate_algo_balance(account_info, amount)
    else:
        decimals = get_asset_decimals(asset_id, algod_client)
        _validate_asset_balance(account_info, asset_id, decimals, amount)


def validate_address(address: str) -> None:
    """
    Check if a given address is a valid Algorand account address.

    Args:
        address (str): The address to be validated.

    Raises:
        click.ClickException: If the address is invalid.

    Returns:
        None
    """

    if not algosdk.encoding.is_valid_address(address):  # type: ignore[no-untyped-call]
        raise click.ClickException(f"`{address}` is an invalid account address")


def get_account_with_private_key(address: str) -> Account:
    """
    Retrieves an account object with the private key based on the provided address.

    Args:
        address (str): The address for which to retrieve the account object.

    Returns:
        Account: An account object with the address and private key.

    Raises:
        click.ClickException: If the address is not valid or if the alias does not exist or does not have a private key.
    """

    parsed_address = address.strip('"')

    try:
        validate_address(parsed_address)
        pk = get_private_key_from_mnemonic()
        return Account(address=parsed_address, private_key=pk)
    except click.ClickException as ex:
        alias_data = get_alias(parsed_address)

        if not alias_data:
            raise click.ClickException(f"Alias `{parsed_address}` alias does not exist.") from ex
        if not alias_data.private_key:
            raise click.ClickException(f"Alias `{parsed_address}` does not have a private key.") from ex

        return Account(address=alias_data.address, private_key=alias_data.private_key)


def get_address(address: str) -> str:
    """
    Validates the given address and returns it if valid. If the address is not valid,
    it checks if the address is an alias and returns the corresponding address if the alias exists.

    Args:
        address (str): The address to be validated or checked for an alias.

    Returns:
        str: The validated address or the corresponding address from an alias.

    Raises:
        click.ClickException: If the address is not valid and no corresponding alias exists.

    Example:
        address = get_address("ABCD1234")
        print(address)
    """

    parsed_address = address.strip('"')

    try:
        validate_address(parsed_address)
        return parsed_address
    except click.ClickException as ex:
        if len(parsed_address) == algosdk.constants.address_len:
            raise click.ClickException(f"`{parsed_address}` is an invalid account address") from ex

        alias_data = get_alias(parsed_address)

        if not alias_data:
            raise click.ClickException(f"Alias `{parsed_address}` alias does not exist.") from ex

        return alias_data.address


def stdin_has_content() -> bool:
    """
    Checks if there is content in the standard input.

    Returns:
        bool: True if there is content in the standard input, False otherwise.
    """

    mode = os.fstat(sys.stdin.fileno()).st_mode
    return stat.S_ISFIFO(mode) or stat.S_ISREG(mode)


def validate_account_balance_to_opt_in(
    algod_client: algosdk.v2client.algod.AlgodClient, account: Account, num_assets: int
) -> None:
    """
    Validates the balance of an account before opt in operation.
    Each asset requires 0.1 Algos to opt in.

    Args:
        algod_client (algosdk.v2client.algod.AlgodClient): The AlgodClient object for
        interacting with the Algorand blockchain.
        account (Account | str): The account object.
        num_assets (int): The number of the assets for opt in (0 for Algos).

    Raises:
        click.ClickException: If there is an insufficient fund in the account or account is not valid.
    """

    address = account.address if isinstance(account, Account) else account
    account_info = algod_client.account_info(address)

    if not isinstance(account_info, dict):
        raise click.ClickException("Invalid account info response")

    required_microalgos = num_assets * algos_to_microalgos(0.1)  # type: ignore[no-untyped-call]
    available_microalgos = account_info.get("amount", 0)
    if available_microalgos < required_microalgos:
        required_algo = microalgos_to_algos(required_microalgos)  # type: ignore[no-untyped-call]
        available_algos = microalgos_to_algos(available_microalgos)  # type: ignore[no-untyped-call]
        raise click.ClickException(
            f"Insufficient Algos balance in account to opt in, required: {required_algo} Algos, available:"
            f" {available_algos} Algos"
        )


def get_account_info(algod_client: algosdk.v2client.algod.AlgodClient, account_address: str) -> dict:
    account_info = algod_client.account_info(account_address)
    assert isinstance(account_info, dict)
    return account_info
