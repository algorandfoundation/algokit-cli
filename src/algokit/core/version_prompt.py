import logging
import os
import re
from datetime import timedelta
from time import time

import click
import httpx

from algokit.core.conf import get_app_config_dir, get_app_state_dir, get_current_package_version

logger = logging.getLogger(__name__)

LATEST_URL = "https://api.github.com/repos/algorandfoundation/algokit-cli/releases/latest"
VERSION_CHECK_INTERVAL = timedelta(weeks=1).total_seconds()
DISABLE_CHECK_MARKER = "disable-version-prompt"


def do_version_prompt() -> None:
    if _skip_version_prompt():
        logger.debug("Version prompt disabled")
        return

    current_version = get_current_package_version()
    latest_version = get_latest_version_or_cached()
    if latest_version is None:
        logger.debug("Could not determine latest version")
        return

    if _get_version_sequence(current_version) < _get_version_sequence(latest_version):
        logger.info(f"You are using AlgoKit version {current_version}, however version {latest_version} is available.")
    else:
        logger.debug("Current version is up to date")


def _get_version_sequence(version: str) -> list[int | str]:
    match = re.match(r"(\d+)\.(\d+)\.(\d+)(.*)", version)
    if match:
        return [int(x) for x in match.groups()[:3]] + [match.group(4)]
    return [version]


def get_latest_version_or_cached() -> str | None:
    version_check_path = get_app_state_dir() / "last-version-check"

    try:
        last_checked = os.path.getmtime(version_check_path)
        version = version_check_path.read_text(encoding="utf-8")
    except OSError:
        logger.debug(f"{version_check_path} inaccessible")
        last_checked = 0
        version = None
    else:
        logger.debug(f"{version} found in cache {version_check_path}")

    if (time() - last_checked) > VERSION_CHECK_INTERVAL:
        try:
            version = get_latest_github_version()
        except Exception as ex:
            logger.debug("Checking for latest version failed", exc_info=ex)
            # update last checked time even if check failed
            version_check_path.touch()
        else:
            version_check_path.write_text(version, encoding="utf-8")
    # handle case where the first check failed, so we have an empty file
    return version or None


def get_latest_github_version() -> str:
    headers = {"ACCEPT": "application/vnd.github+json", "X-GitHub-Api-Version": "2022-11-28"}

    response = httpx.get(LATEST_URL, headers=headers)
    response.raise_for_status()

    json = response.json()
    tag_name = json["tag_name"]
    logger.debug(f"Latest version tag: {tag_name}")
    match = re.match(r"v(\d+\.\d+\.\d+)", tag_name)
    if not match:
        raise ValueError(f"Unable to extract version from tag_name: {tag_name}")
    return match.group(1)


def _skip_version_prompt() -> bool:
    disable_marker = get_app_config_dir() / DISABLE_CHECK_MARKER
    return disable_marker.exists()


skip_version_check_option = click.option(
    "--skip-version-check",
    is_flag=True,
    show_default=False,
    default=False,
    help="Skip version checking and prompting.",
)


@click.command(
    "version-prompt", short_help="Enables or disables checking and prompting if a new version of AlgoKit is available"
)
@click.argument("enable", required=False, type=click.Choice(["enable", "disable"]), default=None)
def version_prompt_configuration_command(*, enable: str | None) -> None:
    """Controls whether AlgoKit checks and prompts for new versions.
    Set to [disable] to prevent AlgoKit performing this check permanently, or [enable] to resume checking.
    If no argument is provided then outputs current setting.

    Also see --skip-version-check which can be used to disable check for a single command."""
    if enable is None:
        logger.info("disable" if _skip_version_prompt() else "enable")
    else:
        disable_marker = get_app_config_dir() / DISABLE_CHECK_MARKER
        if enable == "enable":
            disable_marker.unlink(missing_ok=True)
            logger.info("ðŸ“¡ Resuming check for new versions")
        else:
            disable_marker.touch()
            logger.info("ðŸš« Will stop checking for new versions")
