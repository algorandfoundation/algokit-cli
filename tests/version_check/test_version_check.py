import os
from importlib import metadata
from time import time

import pytest
from algokit.core.conf import PACKAGE_NAME
from algokit.core.version_prompt import LATEST_URL, VERSION_CHECK_INTERVAL
from approvaltests.scrubbers.scrubbers import Scrubber, combine_scrubbers
from pytest_httpx import HTTPXMock
from pytest_mock import MockerFixture

from tests.utils.app_dir_mock import AppDirs
from tests.utils.approvals import normalize_path, verify
from tests.utils.click_invoker import invoke

CURRENT_VERSION = metadata.version(PACKAGE_NAME)
NEW_VERSION = "999.99.99"


def make_scrubber(app_dir_mock: AppDirs) -> Scrubber:
    return combine_scrubbers(
        lambda x: normalize_path(x, str(app_dir_mock.app_config_dir), "{app_config}"),
        lambda x: normalize_path(x, str(app_dir_mock.app_state_dir), "{app_state}"),
        lambda x: x.replace(CURRENT_VERSION, "{current_version}"),
        lambda x: x.replace(NEW_VERSION, "{new_version}"),
    )


@pytest.fixture(autouse=True)
def _setup(mocker: MockerFixture, app_dir_mock: AppDirs) -> None:
    mocker.patch("algokit.core.version_prompt.get_app_config_dir").return_value = app_dir_mock.app_config_dir
    mocker.patch("algokit.core.version_prompt.get_app_state_dir").return_value = app_dir_mock.app_state_dir
    # make bootstrap env a no-op
    mocker.patch("algokit.cli.bootstrap.bootstrap_env")


def test_version_check_queries_github_when_no_cache(app_dir_mock: AppDirs, httpx_mock: HTTPXMock) -> None:
    httpx_mock.add_response(url=LATEST_URL, json={"tag_name": f"v{NEW_VERSION}"})

    # bootstrap env is a nice simple command we can use to test the version check side effects
    result = invoke("bootstrap env", skip_version_check=False)

    assert result.exit_code == 0
    verify(result.output, scrubber=make_scrubber(app_dir_mock))


@pytest.mark.parametrize(
    ("current_version", "latest_version", "warning_expected"),
    [
        ("0.2.0", "0.3.0", True),
        ("0.25.0", "0.30.0", True),
        ("0.3.0", "0.29.0", True),
        ("999.99.99", "1000.00.00", True),
        ("999.99.99-beta", "1000.00.00", True),
        ("999.99.99-alpha", "999.99.99-beta", True),
        ("0.25.0", "1.0.0", True),
        ("0.29.0", "1.0.0", True),
        ("0.3.0", "1.0.0", True),
        ("0.3.0", "0.2.0", False),
        ("0.3.0", "0.3.0", False),
        ("0.30.0", "0.25.0", False),
        ("0.29.0", "0.3.0", False),
        ("0.30.0", "0.30.0", False),
        ("1.0.0", "0.25.0", False),
        ("1.0.0", "0.29.0", False),
        ("1.0.0", "0.3.0", False),
        ("1.0.0", "1.0.0", False),
        ("999.99.99", "998.0.0", False),
        ("999.99.99", "999.99.0", False),
        ("999.99.99", "999.99.99", False),
        ("999.99.99-beta", "998.99.99", False),
        ("999.99.99-beta", "999.99.99-alpha", False),
    ],
)
def test_version_check_only_warns_if_newer_version_is_found(
    app_dir_mock: AppDirs, mocker: MockerFixture, current_version: str, latest_version: str, *, warning_expected: bool
) -> None:
    mocker.patch("algokit.core.version_prompt.get_current_package_version").return_value = current_version
    version_cache = app_dir_mock.app_state_dir / "last-version-check"
    version_cache.write_text(latest_version, encoding="utf-8")
    result = invoke("bootstrap env", skip_version_check=False)

    if warning_expected:
        assert f"version {latest_version} is available" in result.output
    else:
        assert f"version {latest_version} is available" not in result.output


def test_version_check_uses_cache(app_dir_mock: AppDirs) -> None:
    version_cache = app_dir_mock.app_state_dir / "last-version-check"
    version_cache.write_text("1234.56.78", encoding="utf-8")
    result = invoke("bootstrap env", skip_version_check=False)

    assert result.exit_code == 0
    verify(result.output, scrubber=make_scrubber(app_dir_mock))


def test_version_check_queries_github_when_cache_out_of_date(app_dir_mock: AppDirs, httpx_mock: HTTPXMock) -> None:
    httpx_mock.add_response(url=LATEST_URL, json={"tag_name": f"v{NEW_VERSION}"})
    version_cache = app_dir_mock.app_state_dir / "last-version-check"
    version_cache.write_text("1234.56.78", encoding="utf-8")
    modified_time = time() - VERSION_CHECK_INTERVAL - 1
    os.utime(version_cache, (modified_time, modified_time))

    result = invoke("bootstrap env", skip_version_check=False)

    assert result.exit_code == 0
    verify(result.output, scrubber=make_scrubber(app_dir_mock))


def test_version_check_respects_disable_config(app_dir_mock: AppDirs) -> None:
    (app_dir_mock.app_config_dir / "disable-version-prompt").touch()
    result = invoke("bootstrap env", skip_version_check=False)

    assert result.exit_code == 0
    verify(result.output, scrubber=make_scrubber(app_dir_mock))


def test_version_check_respects_skip_option(app_dir_mock: AppDirs) -> None:
    result = invoke("--skip-version-check bootstrap env", skip_version_check=False)

    assert result.exit_code == 0
    verify(result.output, scrubber=make_scrubber(app_dir_mock))


def test_version_check_disable_version_check(app_dir_mock: AppDirs) -> None:
    disable_version_check_path = app_dir_mock.app_config_dir / "disable-version-prompt"
    result = invoke("config version-prompt disable")

    assert result.exit_code == 0
    assert disable_version_check_path.exists()
    verify(result.output, scrubber=make_scrubber(app_dir_mock))


def test_version_check_enable_version_check(app_dir_mock: AppDirs) -> None:
    disable_version_check_path = app_dir_mock.app_config_dir / "disable-version-prompt"
    disable_version_check_path.touch()
    result = invoke("config version-prompt enable")

    assert result.exit_code == 0
    assert not disable_version_check_path.exists()
    verify(result.output, scrubber=make_scrubber(app_dir_mock))


@pytest.mark.parametrize(
    ("method", "message"),
    [
        ("snap", "snap refresh algokit"),
        ("brew", "brew upgrade algokit"),
        ("winget", "winget upgrade algokit"),
        (None, "the tool used to install AlgoKit"),
    ],
)
def test_version_prompt_according_to_distribution_method(
    mocker: MockerFixture, app_dir_mock: AppDirs, method: str, message: str
) -> None:
    mocker.patch("algokit.core.version_prompt._get_distribution_method").return_value = method
    mocker.patch("algokit.core.version_prompt.is_binary_mode").return_value = True
    mocker.patch("algokit.core.version_prompt.get_current_package_version").return_value = "1.0.0"
    version_cache = app_dir_mock.app_state_dir / "last-version-check"
    version_cache.write_text("2.0.0", encoding="utf-8")

    result = invoke("bootstrap env", skip_version_check=False)
    assert result.exit_code == 0
    assert message in result.output
