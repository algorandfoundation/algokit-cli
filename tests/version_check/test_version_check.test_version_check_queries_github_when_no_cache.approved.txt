DEBUG: {app_state}/last-version-check inaccessible
DEBUG: HTTP Request: GET https://api.github.com/repos/algorandfoundation/algokit-cli/releases/latest "HTTP/1.1 200 OK"
DEBUG: Latest version tag: v{new_version}
You are using AlgoKit version {current_version}, however version {new_version} is available.
DEBUG: No .algokit.toml file found in the project directory.
